rts/motoko-rts/src/gc/mark_compact.rs

```diff
@@ -123,87 +126,89 @@ unsafe fn mark_object<M: Memory>(mem: &mut M, obj: SkewedPtr, heap_base: u32) {
-    let mut free = heap_base;
-
-    let mut bitmap_iter = iter_bits();
-    let mut bit = bitmap_iter.next();
-    while bit != BITMAP_ITER_END {
-        let p = (heap_base + (bit * WORD_SIZE)) as *mut Obj;
-
-        // Update forward references to the object to the object's new location and restore
-        // object header
-        unthread(p, free);
-
-        // Get the size before threading the fields, to handle self references.
-        let size = object_size(p as usize).to_bytes().0;
-
-        // Thread fields
-        thread_obj_fields(p, heap_base);
-
-        free += size;
-
-        bit = bitmap_iter.next();
+/// Specialized version of `mark_fields` for root `MutBox`es.
+unsafe fn mark_root_mutbox_fields<M: Memory>(mem: &mut M, mutbox: *mut MutBox, heap_base: u32) {
+    let field_addr = &mut (*mutbox).field;
+    // TODO: Not sure if this check is necessary?
+    if pointer_to_dynamic_heap(field_addr, heap_base as usize) {
+        // TODO: We should be able to omit the "already marked" check here as no two root MutBox
+        // can point to the same object (I think)
+        mark_object(mem, *field_addr, heap_base);
+        // It's OK to thread forward pointers here as the static objects won't be moved, so we will
+        // be able to unthread objects pointed by these fields later.
+        thread(field_addr);
     }
 }
 
-/// Expects all fields to be threaded. Updates backward references and moves objects to their new
-/// locations.
-unsafe fn update_bwd_refs<SetHp: Fn(u32)>(set_hp: SetHp, heap_base: u32) {
+/// Linearly scan the heap, for each live object:
+///
+/// - Mark step threads all backwards pointers and pointers from roots, so unthread to update those
+///   pointers to the objects new location.
+///
+/// - Move the object
+///
+/// - Thread forward pointers of the object
+///
+unsafe fn update_refs<SetHp: Fn(u32)>(set_hp: SetHp, heap_base: u32) {
     let mut free = heap_base;
 
     let mut bitmap_iter = iter_bits();
     let mut bit = bitmap_iter.next();
     while bit != BITMAP_ITER_END {
         let p = (heap_base + (bit * WORD_SIZE)) as *mut Obj;
+        let p_new = free;
 
-        // Update backward references to the object's new location and restore object header
-        unthread(p, free);
+        // Update backwards references to the object's new location and restore object header
+        unthread(p, p_new);
 
-        // All references to the object now point to the new location, move the object
+        // Move the object
         let p_size_words = object_size(p as usize);
-        if free as usize != p as usize {
-            memcpy_words(free as usize, p as usize, p_size_words);
+        if p_new as usize != p as usize {
+            memcpy_words(p_new as usize, p as usize, p_size_words);
         }
 
         free += p_size_words.to_bytes().0;
 
+        // Thread forward pointers of the object
+        thread_fwd_pointers(p_new as *mut Obj, heap_base);
```

Comment:
Is this really the right order in which to do this?
Something tells me you should be unthreading the forward pointers before doing the field move, as the test for whether something is a forward pointer depends on the address of the field (or original object).
Can you convince me this is correct?

Reply:
> as the test for whether something is a forward pointer depends on the address of the field

Two things:

- This algorithm never reorders objects, so if A has a forward pointer to B, then after sliding A it will still have a forward pointer to B. I think this may be more clear if you consider that we only do linear scan from start of the heap to the end, and slide an object at an earlier address before objects at later addresses. So a forward pointer never becomes a backward pointer (or the other way around) as a result of moving objects.

- Threading and then moving would not work. To see why, suppose I have this heap: (assume that we store pointers directly, without skewing)
  ```
  0x100: [ARRAY, ..., 0x200]
  ...
  0x200: [BLOB, ...]
  ```
  So object at 0x100 has a forward pointer to object at 0x200.

  If I thread fields of 0x100 now (before sliding it) it will look like this:

  ```
  0x100: [ARRAY, ..., BLOB]
  ...
  0x200: [0x100, ...]
  ```
  Now if I slide object at `0x100` to its new location:
  ```
  0x050: [ARRAY, ..., BLOB]
  ...
  0x200: [0x100, ...]
  ```
  Note that I have a pointer to random data in the header slot of `0x200` now. I can't unthread it anymore.

  If I move first, then thread, it looks like this:
  ```
  0x050: [ARRAY, ..., BLOB]
  ...
  0x200: [0x050, ...]
  ```
  Now when I visit object at 0x200 I can unthread it.

Does this make sense?

Reply:
Yes, that makes sense. I was being stupid.

Reply:
You're diagrams are always excellent btw